key_file3.txt

```diff
@@ -1,197 +1,189 @@
-3:  6 777 88888 999999 
-4:  5 666 77777 888888 9999999 
-5:  4 555 66666 777777 8888888 99999999 
-6:  3 444 55555 666666 7777777 88888888 999999999 
-7:  2 333 
-8:  1 
-
-Remove 666666 True
-Remove 777777 True
-Remove 888888 True
-Remove 999999 True
-PrintBST
-0:  9 
-1:  8 999 
-2:  7 888 99999 
-3:  6 777 88888 9999999 
-4:  5 666 77777 8888888 99999999 
-5:  4 555 66666 7777777 88888888 999999999 
-6:  3 444 55555 
-7:  2 333 
-8:  1 
-
-Remove 88888888 True
-Remove 99999999 True
-PrintBST
-0:  9 
-1:  8 999 
-2:  7 888 99999 
-3:  6 777 88888 9999999 
-4:  5 666 77777 8888888 999999999 
-5:  4 555 66666 7777777 
-6:  3 444 55555 
-7:  2 333 
-8:  1 
-
-Remove 999999999 True
-PrintBST
-0:  9 
-1:  8 999 
-2:  7 888 99999 
-3:  6 777 88888 9999999 
-4:  5 666 77777 8888888 
-5:  4 555 66666 7777777 
-6:  3 444 55555 
-7:  2 333 
-8:  1 
-
-Remove 7777777 True
-Remove 8888888 True
-Remove 9999999 True
-PrintBST
-0:  9 
-1:  8 999 
-2:  7 888 99999 
-3:  6 777 88888 
-4:  5 666 77777 
-5:  4 555 66666 
-6:  3 444 55555 
-7:  2 333 
-8:  1 
-
-Remove 55555 True
-Remove 66666 True
-Remove 77777 True
-Remove 88888 True
-Remove 99999 True
-PrintBST
-0:  9 
-1:  8 999 
-2:  7 888 
-3:  6 777 
-4:  5 666 
-5:  4 555 
-6:  3 444 
-7:  2 333 
-8:  1 
-
-Remove 333 True
-Remove 444 True
-Remove 555 True
-Remove 666 True
-Remove 777 True
-Remove 888 True
-Remove 999 True
-PrintBST
-0:  9 
-1:  8 
-2:  7 
-3:  6 
-4:  5 
-5:  4 
-6:  3 
-7:  2 
-8:  1 
-
-Remove 1 True
-Remove 2 True
-Remove 3 True
-Remove 4 True
-Remove 5 True
-Remove 6 True
-Remove 7 True
-Remove 8 True
-Remove 9 True
-PrintBST
-BST is empty
-Clear
-PrintBST
-BST is empty
+--- Part 3 output ---
+
+Pretty printing the tree...
+
+Empty tree
+
+Inserting 8, 4, 12, 2, 6, 10, 14, 1, 3, 5, 7,
+          9, 11, 13, and 15 into the tree...
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   2       6      10      14
+
+ 1   3   5   7   9  11  13  15
+
+Attempting to remove 6 from the tree...true
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   2       5      10      14
+
+ 1   3       7   9  11  13  15
+
+Attempting to remove 5 from the tree...true
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   2       7      10      14
+
+ 1   3           9  11  13  15
+
+Attempting to remove 7 from the tree...true
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   2              10      14
+
+ 1   3           9  11  13  15
+
+Attempting to remove 2 from the tree...true
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   1              10      14
+
+     3           9  11  13  15
+
+Attempting to remove 1 from the tree...true
+Pretty printing the tree...
+
+               8
+
+       4              12
+
+   3              10      14
+
+                 9  11  13  15
+
+Attempting to remove 9 from the tree...true
+Pretty printing the tree...
```

Comment:
Skip

Reply:
Remove 12 at this step. This will test their logic for finding the in-order predecessor.

Reply:
Then test for removing 1, 7, 12, and 18 (which are all absent).

Reply:
Finally, do a quick removal of all the nodes (no pretty printing, just true/false; just repeatedly remove the root) to see that they handle the empty tree scenario correctly.

Reply:
I'm a bit confused by your feedback on the tests, as 6 (the first remove) isn't a leaf and if we skip removing 1 and 7, they won't be absent as you stated above. I wonder if reworking the tests would be best. I propose the following:

Start with the original tree that looks like this:
```
               8

       4              12

   2       6      10      14

 1   3   5   7   9  11  13  15
```

Then remove the following nodes:
5 -- leaf node
6 -- node with only a right child
3 -- another leaf node, but it'll set up for the next remove
2 -- node with only a left child
4 -- node with two leaf children
12 -- testing the logic of finding an inorder predecessor

After having hit those cases, test some absent values:
3
12
0
18

Finally, repeatedly remove the head until empty:
8
7
1
11
10
9
14
13
15